File: 161 One Edit Distance.js
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isOneEditDistance = function(s, t) {
    var ls = s.length;
    var lt = t.length;
    if(ls === 0) return lt === 1;
    if(lt === 0) return ls === 1;
    if(lt - ls > 1 || ls - lt > 1) return false;

    if(ls > lt){
       var m = s;
       s = t;
       t = m;
    }
    
    ls = s.length;
    lt = t.length;
    var i,j;
    if(lt - ls === 1){
        for(i = 0,j = 0;i < ls && j - i < 2;i++,j++){
            if(s[i] !== t[j]){
                i--;
            }
        }
        console.log(i +", "+j);
        return j - i < 2;
    }
    
    m = 0;
    for(i = 0;i< ls && m < 2;i++){
        if(s[i] !== t[i]) m++;
    }
    return m === 1;
};

File: 161 One Edit Distance.py
'''
161. One Edit Distance
https://leetcode.com/problems/one-edit-distance/

Given two strings s and t, determine if they are both one edit distance apart.

Note: 

There are 3 possiblities to satisify one edit distance apart:

Insert a character into s to get t
Delete a character from s to get t
Replace a character of s to get t
Example 1:

Input: s = "ab", t = "acb"
Output: true
Explanation: We can insert 'c' into s to get t.
Example 2:

Input: s = "cab", t = "ad"
Output: false
Explanation: We cannot get t from s by only one step.
Example 3:

Input: s = "1203", t = "1213"
Output: true
Explanation: We can replace '0' with '1' to get t.
'''
class Solution:
    def isOneEditDistance(self, s: str, t: str) -> bool:
        # Ensure s is smaller string
        if len(s) > len(t):
            s, t = t, s
        
        ns = len(s)
        nt = len(t)
        if nt - ns > 1:
            return False
        
        for i in range(len(s)):
            if s[i] != t[i]:
                if len(s) == len(t):
                    return s[i+1:] == t[i+1:]
                else: # len(s) + 1 == len(t)
                    return s[i:] == t[i+1:]
        
        return nt - ns == 1
